### DISCUSSION TITLE
What is the best way to implement a Markdown field?

### DISCUSSION THREAD

**Discussion by josephdpurcell** [directus/directus#16762] [Archive]
The features of autocapitalize, autocorrect, and autopunctuate are not supported on iOS using the current Markdown field, see https://github.com/directus/directus/issues/16690.

I can get close to the behavior by modifying the code slightly, but it disables the iOS select menu that lets you do things like paste text. So, it's not as simple as making my own interface because I would need to get the CodeMirror5 config to work and there are issues there, see also https://github.com/codemirror/codemirror5/issues/7009.

There is a good possibility that simply upgrading to CodeMirror6 could address these issues. But, that's not a guarantee -- some investigation and testing would need to be done. And the upgrade path may not be trivial, see https://github.com/directus/directus/issues/16061.

I've been digging into this for many hours and days so I thought I would submit a discussion to see if anyone has a suggestion for how to create a good Markdown editing experience within Directus. (Certainly, I could just build a layer on top of Directus with my own editing experience, but I'm hoping to solve for the issue within Directus.)

Some key criteria I'm looking for, including some from comments below:

* Mobile friendly! The Directus admin is excellent on a mobile device, and I am hoping for long-form content authoring within a Markdown field to be just as pleasant.
* Support for these features that are supported by a plain `textarea` input on iOS:
  * Autocapitalization: The first character ever typed, a character after a punctuation, a character at the beginning of a new line each have the caps key enabled so the character is typed as an uppercase character.
  * Autocorrect: Any phrase in the input is a candidate for being autocorrected.
  * Autopunctuation: A double-space after the characters [a-zA-Z0-9] turns into a period followed by 1 space.
* Support for a select menu where when you select a word or words a menu appears with things like Select All, Paste, etc
* The experience should be "good" (for some definition of good) for non-technical users, see also https://github.com/directus/directus/issues/16690#issuecomment-1335918841 which goes into the philosophy/motivations behind Markdown itself
* Ability to insert media from Directus with correctly formatted markdown, eg upload an image, document, video.
* Match the Directus design. The current Markdown field does this well.
* Add minimal weight to Directus FE. The Markdown field currently uses the same lib as the Code field.

The best comparison I've found so far is GitHub's own website. The input fields for comments and tickets use GitHub Flavored Markdown which is a very good experience (except for the autocapitalization feature being a little finicky on new lines).

Here is a spreadsheet that is an attempt to roll up this discussion into a nice view for comparison: https://docs.google.com/spreadsheets/d/18v2G44v0ix41xbAseOIcMcFOMNluc67IJz9f2U5IpRo/edit?usp=sharing

Any and all thoughts appreciated! You can also message me on Directus Discord.

Edit 12/9 9a ET: Add qualifications for UI and weight.
Edit 19/9 12:15p ET: Note re upgrade path to CM6.
Edit 12/18 1p ET: Add spreadsheet.

**Comment by josephdpurcell** [comment]
sindresorhus' [awesome list](https://github.com/sindresorhus/awesome#front-end-development) led me to this great resource for doing some investigation: https://github.com/dok/awesome-text-editing.

**Reply by josephdpurcell** [reply]
Here is a quick list of Markdown editors that all support autocapitalization, autocorrect, and autopunctuation exactly as I expected on iOS (I tested each for this criteria):

* https://marked.js.org/demo/
* https://spec.commonmark.org/dingus/
* http://demo.showdownjs.com/
* https://markdown-it.github.io/
* https://code.google.com/p/pagedown/wiki/PageDown (no demo available?)

Next, I could look into how easily they could be integrated into Directus' interface, notably for media uploads.

**Reply by josephdpurcell** [reply]
Here are some notes on those projects:

* https://github.com/markedjs/marked 28.8k stars, last edit 2 days ago, 32 issues, 4 PRs; MIT license; dedicatd docs site (not versioned?); Contributing info in docs; supports "The original markdown.pl" and the flavors Commonmark and GitHub Flavored Markdown are "Work in progress" which based on the tickets I interpret as "supported with some deficiencies"; no listed users of marked
* https://github.com/commonmark/commonmark.js 1.3k stars, last edit ~1 mo ago, 42 issues, 7 PRs; BSD license?; no docs?; Contributing info in repo; this is first a specification and second a low-level lib not an actual editor? the repos say they are for reference; Discourse, GitHub, GitLab, Reddit, StackExchange are all listed as users, but also Ghost CMS uses it
* https://github.com/showdownjs/showdown 13k stars, last edit 25 days ago, 163 issues, 3 PRs; MIT license; wiki is available but no versioned docs site; Contributing info in repo; supports "original" (John Gruber), "vanilla" is a showdown flavor, and GitHub Flavored Markdown; it says GoogleCloudPlatform, docular, Meteor, md-page use Showdown and Pagedown is a fork of Showdown and Pagedown is used by StackOverflow
* https://github.com/markdown-it/markdown-it 14.6k stars, last edit ~2 mo ago, 10 issues, 5 PRs; MIT license; dedicated docs site for low level API and a dev guide in the repo; supports Commonmark; it does not list who uses markdown-it
* https://github.com/StackExchange/pagedown 385 stars, last edit 2018, 60 issues, 3 PRs; MIT license?; dedicated docs site but minimal; no contributing info, I think its a dead project? Only listed user is StackExchange

I also found while searching:

* https://github.com/benweet/stackedit.js 906 stars, last edit 2018, 22 issues, 0 PRs; MIT license; minimal docs site; no contributing info; based on PageDown; no listed users
* https://github.com/nhn/tui.editor 15.5k stars, last edit Sep 2022, 384 issues, 38 PRs; MIT license; minimal docs site (not versioned); Contributing info in repo; supports Commonmark and GitHub Flavored Markdown; Dooray.com and Unotes VSCode extension are listed as users

**Reply by paescuj** [reply]
Just throwing in a few more:
- https://github.com/Saul-Mirone/milkdown 6.9k stars, last edit 2 minutes ago
- https://github.com/bytedance/bytemd 3.2k stars, last edit Oct 14
- https://github.com/Ionaru/easy-markdown-editor 1.8k stars, last edit Sep 20
- https://github.com/imzbf/md-editor-v3 375 stars, last edit 5 days ago
- https://github.com/barelyhuman/mark 21 stars, last edit Oct 8

**Reply by josephdpurcell** [reply]
Something to consider is evaluating WYSIWYG editors like CodeMirror6, CKEditor, etc. That would open up a lot of popular options, but an extra step would need to taken to identify any incompatibilities between Markdown-specific content authoring and the editor's authoring experience.

**Reply by josephdpurcell** [reply]
The ticket to upgrade to CodeMirror6 (https://github.com/directus/directus/issues/16061) mentions Monaco, and I'll note easy-markdown-editor from the comment below:

* https://github.com/Ionaru/easy-markdown-editor 1.8k stars, last edit ~2 mo ago, 174 issues, 17 PRs; MIT license; no dedicated docs site but readme is detailed; its based on CodeMirror, but previews are rendered via marked in GitHub Flavored Markdown; contributing info in repo; no stated users of it
* https://github.com/microsoft/monaco-editor 32.2k stars, last edit ~1 mo ago, 769 issues, 28 PRs; MIT License; dedicated docs site; its unclear what flavor of markdown is supported, maybe its not opinionated; contributing info in repo; used by VSCode

**Reply by josephdpurcell** [reply]
I found 2 more:

* https://silverbullet.md/, https://github.com/silverbulletmd/silverbullet 819 stars, last edit yesterday, 69 issues, 2 PRs; MIT license; dedicated docs site with demo; based on CodeMirror6; no stated users of it
* https://simplenote.com/, https://github.com/automattic/simplenote-electron: 4.2k stars, last edit  Jan 2022, 211 issues, 10 PRs; GPL-2.0 license; no docs site?; based on react-monaco-editor; developed by Automattic (WordPress)

**Reply by josephdpurcell** [reply]
I'm going to move the comparison list to https://docs.google.com/spreadsheets/d/18v2G44v0ix41xbAseOIcMcFOMNluc67IJz9f2U5IpRo/edit?usp=sharing. If anyone wants to edit, reach out on Discord or some other mechanism.

**Comment by paescuj** [comment]
In general, I'm also in favor of considering to switch to another Markdown editor.
However, I think the decision should not be made solely on the basis of the current lack of support for autocapitalization, etc., assuming that these functionalities are rather used by a minority (see also [80/20 rule](https://docs.directus.io/getting-started/support.html#why-haven-t-you-added-this-feature-or-fixed-that-issue-yet)).
Instead, the main focus should be to get a decent _overall_ user experience.

The upside of the current Markdown field implementation is that it is well integrated into Directus design-wise, so this should be a prerequisite for a new component. Since the current implementation is relying on CodeMirror which is used for the code field anyway, a small drawback might be that another integration of an external library would then have to be maintained.

**Reply by josephdpurcell** [reply]
Yes! Great points. I incorporated them into the description. I vehemently agree with the "overall good experience" for authoring Markdown, which is a bit of a non-statement given Markdown is so general. But I think we can point to examples and say "that's like what good is" for Markdown editing, such as StackExchange and GitHub. Theres a whole matrix of features to consider too, such as a side-by-side rendered view, and whether that is even desirable. This is a big discussion.

**Reply by josephdpurcell** [reply]
Also, the discussion can be limited to "here's one (or many) ways to implement a good Markdown field", for some definition of good.

**Comment by nickreese** [comment]
As a recent Directus convert after maintaining my own dashboard for content, I highly recommend https://github.com/Ionaru/easy-markdown-editor. I tried most of what is on the list above and overall it hit the perfect mix of WYSIWYG for non-techies and still being markdown.

**Reply by josephdpurcell** [reply]
I see it uses CodeMirror5 and the demo has the same autocorrect, autopunctuation, and autocapitalization issues that CodeMirror5 itself has. There is a ticket open to do a full rewrite citing various issues and maintainability https://github.com/Ionaru/easy-markdown-editor/issues/447.

That said, I REALLY like the experience, specifically the way it is WYSIYG formatted (headings are headings) but preserves the markdown characters like the hash tag. It also looks very extensible with nice documentation. Given that it's on CodeMirror5 it might drop in nicely into current Directus. But, for me personally my hangup is the lack of auto-* support on iOS.

**Comment by josephdpurcell** [comment]
# Introduction

Over the last few weeks I have evaluated various editors in order to find the "best" Markdown editor. Below explains the methodology I used for evaluation, my findings, and a recommendation.

# Methodology

I did not complete an exhaustive evaluation of all editors. I evaluated 22 different editors and 14 CMS's. This list was generated from 3 sources: suggestions in this thread, searching the Internet for keywords like "markdown editor" and "markdown cms", and finding cross-links to comparable editors and CMS's. I looked at factors including Vue support, Markdown support (which flavor, preview availability), input type (eg WYSIWYG vs formatted text vs plain text), activity on the project (stars, open PRs, open tickets, most recent commit), whether contributing is encouraged, license, and whether it was a "wrapper" around some other lower-level editor.

# Findings

The full list of findings can be found here: https://docs.google.com/spreadsheets/d/18v2G44v0ix41xbAseOIcMcFOMNluc67IJz9f2U5IpRo/edit#gid=0.

6 were declined declined for recommendation due to lack of activity or being deprecated, though other factors could also cause it to be declined:

* CodeMirror5 - deprecated by CodeMirror6
* Easy MarkDown Editor - based on CodeMirror5, a big rewrite is pending
* stackedit.js - lack of activity
* StrapDown.js - lack of activity
* TeXMe - lack of activity
* zero-md - lack of activity

5 were declined due to lack of compatibility, though other factors could also cause it to be declined:

* mark - does not appear to be extensible
* react-markdown - no Vue support
* react-md-editor - no Vue support
* React Monaco Editor - no Vue support
* Silver Bullet - this seems to be a "whole package" deal, not something easily dropped into Directus

1 was declined because it requires a license: Typora.

5 were declined because they were single purpose, that is they only supported Markdown. This is important because in order to keep code weight down the editor should be multi-purpose:

* ByteMD
* kg-markdown-html-renderrer
* md-editor-v3
* Showdown.js
* Toast UI

This left 5 for consideration that support Vue, are active projects, and are multipurpose:

* CodeMirror6 - code or Markdown
* Milkdown - Markdown or WYSIWYG
* Monaco - code or Markdown
* ProseMirror - Markdown or WYSIWYG
* Tiptap - Markdown or WYSIWYG

# Recommendation

Further evaluation is needed to identify a sure path forward. I believe there are two general paths to take, Option 1 and Option 2 as clarified below. This deeper evaluation should consider specific features for each of the Markdown, Code, and WYSIWYG fields such as extensibility, autosave, collaborative editing, accessibility, internationalization, syntax highlighting, code completion, undo history, preview vs source toggling, and the overall experience and needs for those respective field types.

## Option 1: Use one editor for Markdown and Code Fields and another for WYSIWYG Field

In this scenario, it probably makes most sense to proceed with using CodeMirror for Markdown and Code fields, but upgrade from CodeMirror5 to CodeMirror6 to address defects. Continue using TinyMCE for WYSIWYG.

The alternative would be to use Monaco instead of CodeMirror for the Markdown and Code fields since it seems (on the surface) to offer comparable functionality, but there are reasons to decide against Monaco. One, Directus already has experience with CodeMirror. Two, of the Markdown editors I found there was only 1 based on Monaco but at least 2 based on CodeMirror. Third, of the CMS's I evaluated none of them used Monaco for the Markdown inputs but 3 of them used CodeMirror.

## Option 2: Use one editor for Markdown and WYSIWYG Fields and another for Code Field

In this scenario, it probably makes sense to use CodeMirror for the Code field. Upgrading from CodeMirror5 to 6 is ideal, but optional in the near term.

TinyMCE would be removed from the project and a ProseMirror editor would be used instead. There are 3 options:

* Milkdown
* ProseMirror
* Tiptap

Further evaluation would need to be done to identify which is the best fit. They all have examples of a WYSIWYG view so they should be compatible for the WYSIWYG field. But, I did not find examples of how to use these editors in a side-by-side view or toggle view which allows the author to write in raw Markdown. However, that should be supported. They are all based on ProseMirror and its documentation shows that a raw view would be supported as shown here: https://prosemirror.net/examples/markdown/.

**Reply by josephdpurcell** [reply]
My gut reaction is that Option 1 is the fast-path to addressing what I personally want to see, which is https://github.com/directus/directus/issues/16690 addressed. Consequently, it would leave TinyMCE in use which has a defect https://github.com/directus/directus/issues/16699; while that may not be mutually exclusive, I'm not certain there is a fix in TinyMCE for it.

But, Option 2 seems the most appealing. It seems like a ProseMirror based editor is a great path to take for both the Markdown and WYSIWYG fields. I keep referring to GitHub's input as a reference for a "good" Markdown authoring experience and based on what I can see GitLab is using Tiptap which is based on ProseMirror, and GitLab's authoring experience is comparable to GitHub's. Going the ProseMirror route comes with the bonus of being able to use it for a WYSIWYG.

Then again, if an upgrade to CodeMirror6 truly is a quick thing to do, perhaps it makes sense to go that route and allow an evaluation of swapping editors out be put on a longer term roadmap where the evaluation can get more detailed, and there's a LOT of detail to wade through (eg should typing "h1. " in a WYSIWYG switch to a heading? should typing "**" in a Markdown field show as bold text or be left as-is? and a thousand more questions)

**Reply by josephdpurcell** [reply]
After consulting my pillow I'm still aligned with my previous comment. I don't see a compelling reason to remove CodeMirror from the project. If that is consensus, then it would be beneficial to proceed with the upgrade to CodeMirror6. Meanwhile further evaluation about functionality of the Markdown, Code, and WYSIWYG fields can progress.

**Reply by josephdpurcell** [reply]
If anyone has time it would be excellent for someone to review the analysis above and see if anything was missed. There could be a compelling reason to use Monaco or some other lib.

**Reply by rijkvanzanten** [reply]
Woa great research! The current markdown editor is certainly not as in-depth as what you've proposed here 🙂 My original MVP was "just" to have a decent monospace editor with some light highlighting and a preview view. A lot of the above examples blur the line between wysiwyg and markdown quite a bit more, which is nice!

> 5 were declined because they were single purpose, that is they only supported Markdown. This is important because in order to keep code weight down the editor should be multi-purpose:

This is an interesting take! When it comes to an admin backend environment as Directus, the overall size of the bundles matters slightly less. Having a larger upfront loading time is acceptable if the editing experience is better. I have yet to come across a single library that does WYSIWYG well in markdown, html, and JSON.

> The alternative would be to use Monaco instead of CodeMirror for the Markdown and Code fields since it seems (on the surface) to offer comparable functionality, but there are reasons to decide against Monaco. One, Directus already has experience with CodeMirror. Two, of the Markdown editors I found there was only 1 based on Monaco but at least 2 based on CodeMirror. Third, of the CMS's I evaluated none of them used Monaco for the Markdown inputs but 3 of them used CodeMirror.

Objection! Monaco is way newer, which would also explain the reason you'd find less libraries relying on it for the editing experience. When it comes to code-editing; monaco is what's driving VS Code, which I'd guess has more use than all codemirror-based markdown editor libraries combined! As for "Directus has experience".. well.. We already concluded CodeMirror 5 has to be replaced, and from what I can tell so far, it's not just a drop-in replacement upgrade. If Monaco is the better choice in an out-of-context head-to-head, I'm very down to discuss using that in Directus over CodeMirror 6.

> TinyMCE would be removed from the project and a ProseMirror editor would be used instead. 

Isn't ProseMirror saving to a custom JSON based format rather than HTML? I'm not too clear on the details, but that's what I vaguely remember from the last time I did a similar research 🙂

**Reply by rijkvanzanten** [reply]
> My gut reaction is that Option 1 is the fast-path to addressing what I personally want to see, which is https://github.com/directus/directus/issues/16690 addressed. Consequently, it would leave TinyMCE in use which has a defect https://github.com/directus/directus/issues/16699; while that may not be mutually exclusive, I'm not certain there is a fix in TinyMCE for it.

I agree. I think there will be way more overlap between markdown-code than markdown-wysiwyg. As a matter of fact, we could render the regular code interface _as part of_ the markdown interface, and wrap it with the Previewing UI. That way we aren't maintaining two different "code" editing instances 🙂
